### PR TITLE
[WIP]Have restructured Position, TokenTup, LexBufState to be more memory &…… copy efficient

### PR DESCRIPTION
**Pull request by gerardtoconnor**
PR for issue #6006 
Looking into deeper and deeper, the allocation time was being spent doing all the coping of large structs like Position, a lot of the Position mutation related to just moving back and forward on the same line so I created a LinePosition Object that is used immutably, but meant Position Struct was now only an Absolute Postion, and a ref to a LinePosition. Now that there is less copying, and state objects are being passed by ref and not mutated, we are going from 8.7% >> 2% for these allocation & copying

(all tests done by going to line 6699 in TypeChecker.fs and erasing tokens (waiting), and replacing back (and waiting)

image

All tests passed except FSharpSuite ones … that never work for me!?
@TIHan @forki

### CODE REVIEW

**Comment by abelbraaksma** [comment]
Wow, impressive improvement! Are the two screenshots interchangeable? I also notice large decreases of `string` and `LargeObject`, are these also a result of this change, or was the sample set different?

**Comment by gerardtoconnor** [comment]
@abelbraaksma I recreated exact same steps in 4 iterations and so would have expected them to be same but have come down for some reason, it's critical someone else tests independently as should be similar

File: src/fsharp/LexFilter.fs [hunk 1]
```diff
@@ -394,26 +394,19 @@ let parenTokensBalance t1 t2 =
...
     member x.StartPos = startPos
     member x.EndPos = endPos
     member x.PastEOF = pastEOF
-
-[<Struct>]
-type PositionTuple =
-    val X: Position
-    val Y: Position
-    new (x: Position, y: Position) = { X = x; Y = y }
-
```
Comment (TIHan) [inline src/fsharp/LexFilter.fs]: After looking at what changed below, getting rid of this is great. `X` was never used, is that correct? So the whole point of `PositionTuple` was just to use `Y`.

File: src/fsharp/LexFilter.fs [hunk 2]
```diff
@@ -868,7 +861,7 @@ type LexFilterImpl (lightSyntaxStatus:LightSyntaxStatus, compilingFsLib, lexer,
                     let lookaheadTokenStartPos = startPosOfTokenTup lookaheadTokenTup
                     match lookaheadToken with 
                     | Parser.EOF _ | SEMICOLON_SEMICOLON -> false 
-                    | _ when indentation && lookaheadTokenStartPos < tokenEndPos -> false
+                    | _ when indentation && lookaheadTokenStartPos.AbsoluteOffset < tokenEndPos.AbsoluteOffset -> false
```
Comment (TIHan) [inline src/fsharp/LexFilter.fs]: Wow, the code before looks very wrong. It was effectively doing a comparison on the type that contained file indices and others, which was probably not intended.

File: src/fsharp/LexFilter.fs [hunk 3]
```diff
@@ -1099,14 +1099,14 @@ type LexFilterImpl (lightSyntaxStatus:LightSyntaxStatus, compilingFsLib, lexer,
...
+                let pos = tokenTup.LastTokenPos
                 pos.ShiftColumnBy 1
-            returnToken (lexbufStateForInsertedDummyTokens (lastTokenPos, tokenTup.LexbufState.StartPos)) tok
+            returnDummyToken lastTokenPos tokenTup.LexbufState.StartPos tok
 
         let insertToken tok = 
             delayToken tokenTup
             if debug then dprintf "inserting %+A\n" tok
-            returnToken (lexbufStateForInsertedDummyTokens (startPosOfTokenTup tokenTup, tokenTup.LexbufState.EndPos)) tok
+            returnDummyToken (startPosOfTokenTup tokenTup) tokenTup.LexbufState.EndPos tok
```
Comment (TIHan) [inline src/fsharp/LexFilter.fs]: This cleans up the code nicely.

**Comment by TIHan** [comment]
@abelbraaksma the `string` and `LargeObject` won't be impacted by this change. The exclusive number seems to be higher in the first screenshot most likely because more stuff probably ran in the trace.

**Comment by TIHan** [comment]
@gerardtoconnor I will test your changes independently to verify perf results. Regardless, I still want to accept the change because it improves the code as well.

File: src/fsharp/LexFilter.fs (hunk 1 above)
Comment (gerardtoconnor) [reply src/fsharp/LexFilter.fs]: Yeah, I noticed that it was only ever being used to find the end position of that lastToken, so the start token position was wasted space. I tested with having only an AbsoluteOffset but due to the fact that these carry over lines, you need the LinePosition data also.

File: src/fsharp/LexFilter.fs [hunk 4]
```diff
@@ -422,9 +415,7 @@ type TokenTup =
     member x.EndPos = x.LexbufState.EndPos
     
     /// Returns a token 'tok' with the same position as this token
-    member x.UseLocation(tok) = 
-        let tokState = x.LexbufState 
-        TokenTup(tok,LexbufState(tokState.StartPos, tokState.EndPos,false),x.LastTokenPos)
+    member x.UseLocation(tok) = TokenTup(tok,x.LexbufState,x.LastTokenPos)
```
[the remark this answers is not in the document]
Comment (gerardtoconnor) [reply src/fsharp/LexFilter.fs]: When looking at this, it felt like a bug, why would the EOF state change at the same location for a copy of token, I switched this in and everything passed on tests so felt was ok to integrate as saved the allocation, Most of the time it will not be EOF (false) so I can do a further commit where it checks if EOF, and if not, it will reuse state, otherwise, create a new one with the false flag?

File: src/fsharp/LexFilter.fs [hunk 5]
```diff
@@ -553,17 +542,21 @@ type LexFilterImpl (lightSyntaxStatus:LightSyntaxStatus, compilingFsLib, lexer,
...
-        setLexbufState state
-        let lastTokenStart = state.StartPos
-        let lastTokenEnd = state.EndPos
+        
+        if haveLexbufState then 
+            setLexbufState savedLexbufState
+        else
+            savedLexbufState <- getLexbufState()
+         
+        let lastTokenEnd = savedLexbufState.EndPos
```
[the remark this answers is not in the document]
Comment (gerardtoconnor) [reply src/fsharp/LexFilter.fs]: Yeah, this code didn't make sense to me, was doing pointless operations so cleaned it up, `lexer` function is mutating the lexbuffer positions and this takes that under consideration … I had a few null reference exceptions in its evolution

File: src/fsharp/LexFilter.fs (hunk 3 above)
Comment (gerardtoconnor) [reply src/fsharp/LexFilter.fs]: We were allocating a lexBufState just to allocate the lexbuffer positions, made no sense, this seemed tidier

File: .gitignore [hunk 6]
```diff
@@ -130,3 +130,10 @@ tests/fsharpqa/testenv/bin/System.ValueTuple.dll
 msbuild.binlog
 /fcs/FSharp.Compiler.Service.netstandard/*.fs
 /fcs/FSharp.Compiler.Service.netstandard/*.fsi
+/FSharp.Core.dll
```
Comment (cartermp) [inline .gitignore]: Just to aid in traceability, can this be factored into a separate PR?

Comment (cartermp) [reply .gitignore]: For context, we often end up going back in time to see what else was changed around a particular line of code and get confused by seemingly unrelated changes.

File: src/fsharp/LexFilter.fs (hunk 4 above)
[the remark this answers is not in the document]
Comment (TIHan) [reply src/fsharp/LexFilter.fs]: Yea, most of the time it will be EOF.

I think we should put in the check just in case. Also, put a comment describing why we are doing it.

File: .gitignore (hunk 6 above)
Comment (gerardtoconnor) [reply .gitignore]: This was unintentional, I have rolled back to remove these, good spot

File: tests/FSharp.Core.UnitTests/FSharp.Core.UnitTests.nuget.props [hunk 7]
```diff
@@ -0,0 +1,19 @@
+﻿<?xml version="1.0" encoding="utf-8" standalone="no"?>
```
[the remark this answers is not in the document]
Comment (gerardtoconnor) [reply tests/FSharp.Core.UnitTests/FSharp.Core.UnitTests.nuget.props]: Ugh, snuck in there sorry, removing now …

**Comment by TIHan** [comment]
I haven't had time this evening to verify this. I will get back to it soon @gerardtoconnor . Code changes look good. CI is failing because of duplicate definition:
```
prim-lexing.fsi(37,10): error FS0037: Duplicate definition of field 'AbsoluteOffset'
```

**Comment by gerardtoconnor** [comment]
Cheers GitHub "conflict resolution" , what happens when you edit code outside IDE. Have stripped out the duplicates it added there. Don't worry, there's no rush on the testing Merge … I have a few more PR's in the pipeline conserve your energy 😉

**Comment by TIHan** [comment]
@gerardtoconnor Unfortunately, I don't have good news.

Before your changes:
no_changes

After your changes:
changes

It regresses parsing performance by a significant amount even though we reduced allocations.
I made a PR adding in this specific benchmark: https://github.com/Microsoft/visualfsharp/pull/6017
It tests only TypeChecker.fs. Once this gets merged in I hope you will be able to iterate on it.

I'm glad we caught this early. We need to have good benchmarking and evidence before we start pulling this stuff in.

**Comment by gerardtoconnor** [comment]
No that's fine, the Perfview traces were just trying to improve allocation times, why I wanted to get validated on entire use case. with the new benchmark in, will be easy to tweak till we get the right balance.

**Comment by TIHan** [comment]
Reducing allocations by changing stuff from structs to classes and vice versa, might be able to help; but from this evidence, the best way to approach this is to figure out how to do less work in `LexFilter.fs` itself.

**Comment by gerardtoconnor** [comment]
Will I just close this and I can re-submit once we've figured a way to reduce without regressing perf on benchmark?

**Comment by TIHan** [comment]
I would prefer to keep this open due to discussion and visibility.

If you want to re-do the PR, you can just force push override your branch here.

**Comment by gerardtoconnor** [comment]
I used the benchmark and it gave consistent results with you got, but the strange thing, for me at least, was that I could not find the TokenTup issue in memory running the benchmark while it turns up in VS service?

**Comment by abelbraaksma** [comment]
Since the original issue is about slowing down VS as a result of memory congestion, it's possible that under high memory pressure, this PR would create a noticeable improvement, regardless of a slightly higher pressure on CPU in small projects. Whether the trade off is worth it I don't know.

**Comment by cartermp** [comment]
@gerardtoconnor My guess is because the benchmark is really just measuring one thing and not simulating real-world VS usage, you wouldn't see memory come up much.

@abelbraaksma As a general rule we don't want to regress CPU time perf like this unless the memory usage is massively smaller (i.e., we solved The Issue ™️ that causes all these memory problems). Parsing is used by nearly every feature in the IDE, so slowing that down could make things feel worse for people who do not suffer large memory usage problems in the IDE today.

**Comment by gerardtoconnor** [comment]
Yeah, I'm in agreement, we want to reduce memory if not at cost of perf, there are a few things more we can test to try he both, I'll use combination of perfview and benchmark to find sweet spot

**Comment by gerardtoconnor** [comment]
@cartermp yeah I think we're all in consensus on that, aim is to reduce memory without materially impacting performance, this isn't enough to justify that slow down. I will keep looking for ways to compact the state & minimise allocs to we can get both memory down and same if not better perf in benchmark, As @TIHan already said, will probably also require re-architecting the algo a bit.

@abelbraaksma yeah its not the compromise I wanted to make and hence why was asking for the validation, it's reducing alloc times by 5% but slowing down plain lex case 20% so terrible trade-off. if it was 20% less memory and 5% slowdown it might be worth it so will just need to keep looking

**Comment by KevinRansom** [comment]
@gerardtoconnor I added [WIP], please let us know when you are ready to progress with this.

Thanks,

Kevin

**Comment by gerardtoconnor** [comment]
Would rather restructure the Lexer workflow then do these fringe tidy operations, especially as reduced allocations are not translating into better performance, will close this as regressing in real-world tests.
